solidity/test/maintainer/MaintainerProxy.test.ts

```diff
@@ -0,0 +1,3434 @@
+            authorizedMaintainer.address
+          )
+          tx = await maintainerProxy
+            .connect(authorizedMaintainer)
+            .notifyWalletClosingPeriodElapsed(ecdsaWalletTestData.pubKeyHash160)
+        })
+
+        after(async () => {
+          await walletRegistry.closeWallet.reset()
+
+          await restoreSnapshot()
+        })
+
+        it("should successfully submit sweep proof", async () => {
+          await expect(tx.wait()).not.to.be.reverted
+        })
+
+        it("should refund ETH", async () => {
+          const postNotifyThirdPartyBalance = await provider.getBalance(
+            authorizedMaintainer.address
+          )
+          const diff = postNotifyThirdPartyBalance.sub(initThirdPartyBalance)
+
+          expect(diff).to.be.gt(0)
+          expect(diff).to.be.lt(
+            ethers.utils.parseUnits("1000000", "gwei") // 0,001 ETH
+          )
+        })
+      })
+    })
+  })
+
+  describe("defeatFraudChallengeWithHeartbeat", () => {
+    let heartbeatWalletPublicKey: string
+    let heartbeatWalletSigningKey: SigningKey
+
+    before(async () => {
+      await createSnapshot()
+
+      // For `defeatFraudChallengeWithHeartbeat` unit tests we do not use test
+      // data from `fraud.ts`. Instead, we create random wallet and use its
+      // SigningKey.
+      //
+      // This approach is better long-term. In case the format of the heartbeat
+      // message changes or in case we want to add more unit tests, we can simply
+      // call appropriate function to compute another signature. Also, we do not
+      // use any BTC-specific data for this set of unit tests.
+      const wallet = ethers.Wallet.createRandom()
+      // We use `ethers.utils.SigningKey` for a `Wallet` instead of
+      // `Signer.signMessage` to do not add '\x19Ethereum Signed Message:\n'
+      // prefix to the signed message. The format of the heartbeat message is
+      // the same no matter on which host chain TBTC is deployed.
+      heartbeatWalletSigningKey = new ethers.utils.SigningKey(wallet.privateKey)
+      // Public key obtained as `wallet.publicKey` is an uncompressed key,
+      // prefixed with `0x04`. To compute raw ECDSA key, we need to drop `0x04`.
+      heartbeatWalletPublicKey = `0x${wallet.publicKey.substring(4)}`
+
+      const walletID = keccak256(heartbeatWalletPublicKey)
+      const walletPublicKeyX = `0x${heartbeatWalletPublicKey.substring(2, 66)}`
+      const walletPublicKeyY = `0x${heartbeatWalletPublicKey.substring(66)}`
+      await bridge
+        .connect(walletRegistry.wallet)
+        .__ecdsaWalletCreatedCallback(
+          walletID,
+          walletPublicKeyX,
+          walletPublicKeyY
+        )
+    })
+
+    after(async () => {
+      await restoreSnapshot()
+    })
+
+    context("when the challenge exists", () => {
+      context("when the challenge is open", () => {
+        context("when the heartbeat message has correct format", () => {
```

Comment:
I would consider removing these contexts. No setup happens inside them and they are not relevant for gas refund tests.